File: flynn.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/gorilla/websocket"
)

var (
	socket            *websocket.Conn
	pingTicker        *time.Ticker
	countdownTicker   *time.Ticker
	potentialPoints   float64
	countdown         string
	pointsTotal       int
	pointsToday       int
	localStorageFile  = "localStorage.json"
	reader            = bufio.NewReader(os.Stdin)
)

type LocalStorage struct {
	LastUpdated     string  `json:"lastUpdated"`
	PointsTotal     int     `json:"pointsTotal"`
	PointsToday     int     `json:"pointsToday"`
	LastPingDate    string  `json:"lastPingDate"`
	PotentialPoints float64 `json:"potentialPoints"`
	Countdown       string  `json:"countdown"`
	UserID          string  `json:"userId"`
}

func getLocalStorage() (LocalStorage, error) {
	var data LocalStorage
	fileContent, err := ioutil.ReadFile(localStorageFile)
	if err != nil {
		return data, nil
	}
	err = json.Unmarshal(fileContent, &data)
	return data, err
}

func setLocalStorage(data LocalStorage) error {
	currentData, _ := getLocalStorage()
	newData := currentData

	if data.LastUpdated != "" {
		newData.LastUpdated = data.LastUpdated
	}
	if data.PointsTotal != 0 {
		newData.PointsTotal = data.PointsTotal
	}
	if data.PointsToday != 0 {
		newData.PointsToday = data.PointsToday
	}
	if data.LastPingDate != "" {
		newData.LastPingDate = data.LastPingDate
	}
	if data.PotentialPoints != 0 {
		newData.PotentialPoints = data.PotentialPoints
	}
	if data.Countdown != "" {
		newData.Countdown = data.Countdown
	}
	if data.UserID != "" {
		newData.UserID = data.UserID
	}

	jsonData, err := json.Marshal(newData)
	if err != nil {
		return err
	}
	return ioutil.WriteFile(localStorageFile, jsonData, 0644)
}

func connectWebSocket(userId string) error {
	if socket != nil {
		return nil
	}
	version := "v0.2"
	url := "wss://secure.ws.teneo.pro"
	wsURL := fmt.Sprintf("%s/websocket?userId=%s&version=%s", url, userId, version)

	var err error
	socket, _, err = websocket.DefaultDialer.Dial(wsURL, nil)
	if err != nil {
		return err
	}

	connectionTime := time.Now().UTC().Format(time.RFC3339)
	err = setLocalStorage(LocalStorage{LastUpdated: connectionTime})
	if err != nil {
		return err
	}
	fmt.Println("WebSocket connected at", connectionTime)

	go startPinging()
	go startCountdownAndPoints()

	go func() {
		for {
			_, message, err := socket.ReadMessage()
			if err != nil {
				fmt.Println("WebSocket read error:", err)
				return
			}
			var data map[string]interface{}
			err = json.Unmarshal(message, &data)
			if err != nil {
				fmt.Println("JSON unmarshal error:", err)
				continue
			}
			fmt.Println("Received message from WebSocket:", data)
			if pointsTotal, ok := data["pointsTotal"].(float64); ok {
				if pointsToday, ok := data["pointsToday"].(float64); ok {
					lastUpdated := time.Now().UTC().Format(time.RFC3339)
					err = setLocalStorage(LocalStorage{
						LastUpdated: lastUpdated,
						PointsTotal: int(pointsTotal),
						PointsToday: int(pointsToday),
					})
					if err != nil {
						fmt.Println("Error setting local storage:", err)
					}
				}
			}
		}
	}()

	return nil
}

func disconnectWebSocket() {
	if socket != nil {
		socket.Close()
		socket = nil
		stopPinging()
	}
}

func startPinging() {
	stopPinging()
	pingTicker = time.NewTicker(10 * time.Second)
	go func() {
		for range pingTicker.C {
			if socket != nil {
				err := socket.WriteJSON(map[string]string{"type": "PING"})
				if err != nil {
					fmt.Println("Error sending ping:", err)
					continue
				}
				err = setLocalStorage(LocalStorage{LastPingDate: time.Now().UTC().Format(time.RFC3339)})
				if err != nil {
					fmt.Println("Error setting last ping date:", err)
				}
			}
		}
	}()
}

func stopPinging() {
	if pingTicker != nil {
		pingTicker.Stop()
		pingTicker = nil
	}
}

func startCountdownAndPoints() {
	updateCountdownAndPoints()
	countdownTicker = time.NewTicker(1 * time.Second)
	go func() {
		for range countdownTicker.C {
			updateCountdownAndPoints()
		}
	}()
}

func updateCountdownAndPoints() {
	localStorage, err := getLocalStorage()
	if err != nil {
		fmt.Println("Error getting local storage:", err)
		return
	}

	if localStorage.LastUpdated != "" {
		lastUpdated, err := time.Parse(time.RFC3339, localStorage.LastUpdated)
		if err != nil {
			fmt.Println("Error parsing last updated time:", err)
			return
		}

		nextHeartbeat := lastUpdated.Add(15 * time.Minute)
		now := time.Now().UTC()
		diff := nextHeartbeat.Sub(now)

		if diff > 0 {
			minutes := int(diff.Minutes())
			seconds := int(diff.Seconds()) % 60
			countdown = fmt.Sprintf("%dm %ds", minutes, seconds)

			maxPoints := 25.0
			timeElapsed := now.Sub(lastUpdated)
			timeElapsedMinutes := timeElapsed.Minutes()
			newPoints := math.Min(maxPoints, (timeElapsedMinutes/15)*maxPoints)
			newPoints = math.Floor(newPoints*100) / 100

			if rand.Float64() < 0.1 {
				bonus := rand.Float64() * 2
				newPoints = math.Min(maxPoints, newPoints+bonus)
				newPoints = math.Floor(newPoints*100) / 100
			}

			potentialPoints = newPoints
		} else {
			countdown = "Calculating..."
			potentialPoints = 25
		}
	} else {
		countdown = "Calculating..."
		potentialPoints = 0
	}

	err = setLocalStorage(LocalStorage{
		PotentialPoints: potentialPoints,
		Countdown:       countdown,
	})
	if err != nil {
		fmt.Println("Error setting local storage:", err)
	}
}

func getUserId() error {
	loginURL := "https://ikknngrgxuxgjhplbpey.supabase.co/auth/v1/token?grant_type=password"
	authorization := "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"
	apikey := "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"

	fmt.Print("Email: ")
	email, _ := reader.ReadString('\n')
	email = strings.TrimSpace(email)

	fmt.Print("Password: ")
	password, _ := reader.ReadString('\n')
	password = strings.TrimSpace(password)

	payload := strings.NewReader(fmt.Sprintf(`{"email":"%s","password":"%s"}`, email, password))

	req, err := http.NewRequest("POST", loginURL, payload)
	if err != nil {
		return err
	}

	req.Header.Add("Authorization", authorization)
	req.Header.Add("apikey", apikey)
	req.Header.Add("Content-Type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	var response map[string]interface{}
	err = json.Unmarshal(body, &response)
	if err != nil {
		return err
	}

	user, ok := response["user"].(map[string]interface{})
	if !ok {
		return fmt.Errorf("user data not found in response")
	}

	userId, ok := user["id"].(string)
	if !ok {
		return fmt.Errorf("user ID not found in response")
	}

	fmt.Println("User ID:", userId)

	profileURL := fmt.Sprintf("https://ikknngrgxuxgjhplbpey.supabase.co/rest/v1/profiles?select=personal_code&id=eq.%s", userId)
	req, err = http.NewRequest("GET", profileURL, nil)
	if err != nil {
		return err
	}

	req.Header.Add("Authorization", authorization)
	req.Header.Add("apikey", apikey)

	res, err = http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()

	body, err = ioutil.ReadAll(res.Body)
	if err != nil {
		return err
	}

	fmt.Println("Profile Data:", string(body))

	err = setLocalStorage(LocalStorage{UserID: userId})
	if err != nil {
		return err
	}

	startCountdownAndPoints()
	return connectWebSocket(userId)
}

func main() {
	localStorage, err := getLocalStorage()
	if err != nil {
		fmt.Println("Error getting local storage:", err)
		return
	}

	userId := localStorage.UserID

	if userId == "" {
		fmt.Print("Please enter your user ID: ")
		userId, _ = reader.ReadString('\n')
		userId = strings.TrimSpace(userId)
		err = setLocalStorage(LocalStorage{UserID: userId})
		if err != nil {
			fmt.Println("Error setting user ID:", err)
			return
		}
		startCountdownAndPoints()
		err = connectWebSocket(userId)
		if err != nil {
			fmt.Println("Error connecting to WebSocket:", err)
			return
		}
	} else {
		fmt.Println("Menu:")
		fmt.Println("1. Logout")
		fmt.Println("2. Start Running Node")
		fmt.Println("3. Get User ID")
		fmt.Print("Choose an option: ")

		option, _ := reader.ReadString('\n')
		option = strings.TrimSpace(option)

		switch option {
		case "1":
			err = setLocalStorage(LocalStorage{})
			if err != nil {
				fmt.Println("Error clearing local storage:", err)
			} else {
				fmt.Println("Logged out successfully.")
			}
		case "2":
			startCountdownAndPoints()
			err = connectWebSocket(userId)
			if err != nil {
				fmt.Println("Error connecting to WebSocket:", err)
			}
		case "3":
			err = getUserId()
			if err != nil {
				fmt.Println("Error getting user ID:", err)
			}
		default:
			fmt.Println("Invalid option. Exiting...")
		}
	}

	// Handle SIGINT (Ctrl+C)
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-c
		fmt.Println("Received SIGINT. Stopping pinging...")
		stopPinging()
		disconnectWebSocket()
		os.Exit(0)
	}()

	// Keep the program running
	select {}
}

File: flynn.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <chrono>
#include <thread>
#include <ctime>
#include <iomanip>
#include <random>
#include <nlohmann/json.hpp>
#include <websocketpp/config/asio_client.hpp>
#include <websocketpp/client.hpp>
#include <curl/curl.h>

using json = nlohmann::json;
using websocket_client = websocketpp::client<websocketpp::config::asio_tls_client>;

websocket_client::connection_ptr socket_ptr = nullptr;
std::thread ping_thread;
std::thread countdown_thread;
double potential_points = 0.0;
std::string countdown = "Calculating...";
int points_total = 0;
int points_today = 0;

json get_local_storage() {
    std::ifstream file("localStorage.json");
    if (file.is_open()) {
        json data;
        file >> data;
        return data;
    }
    return json::object();
}

void set_local_storage(const json& data) {
    json current_data = get_local_storage();
    current_data.merge_patch(data);
    std::ofstream file("localStorage.json");
    file << current_data.dump(4);
}

void connect_websocket(const std::string& user_id) {
    if (socket_ptr) return;

    std::string version = "v0.2";
    std::string url = "wss://secure.ws.teneo.pro";
    std::string ws_url = url + "/websocket?userId=" + user_id + "&version=" + version;

    websocket_client client;
    client.init_asio();

    websocketpp::lib::error_code ec;
    socket_ptr = client.get_connection(ws_url, ec);
    if (ec) {
        std::cout << "Could not create connection: " << ec.message() << std::endl;
        return;
    }

    client.connect(socket_ptr);

    client.run();
}

void disconnect_websocket() {
    if (socket_ptr) {
        socket_ptr->close(websocketpp::close::status::normal, "Closing connection");
        socket_ptr = nullptr;
    }
}

void start_pinging() {
    ping_thread = std::thread([&]() {
        while (socket_ptr && socket_ptr->get_state() == websocketpp::session::state::open) {
            json ping_message = {{"type", "PING"}};
            socket_ptr->send(ping_message.dump());
            set_local_storage({{"lastPingDate", std::chrono::system_clock::now().time_since_epoch().count()}});
            std::this_thread::sleep_for(std::chrono::seconds(10));
        }
    });
}

void stop_pinging() {
    if (ping_thread.joinable()) {
        ping_thread.join();
    }
}

void update_countdown_and_points() {
    json local_storage = get_local_storage();
    if (local_storage.contains("lastUpdated")) {
        std::chrono::system_clock::time_point last_updated = std::chrono::system_clock::from_time_t(local_storage["lastUpdated"]);
        auto next_heartbeat = last_updated + std::chrono::minutes(15);
        auto now = std::chrono::system_clock::now();
        auto diff = std::chrono::duration_cast<std::chrono::seconds>(next_heartbeat - now);

        if (diff.count() > 0) {
            int minutes = diff.count() / 60;
            int seconds = diff.count() % 60;
            countdown = std::to_string(minutes) + "m " + std::to_string(seconds) + "s";

            const int max_points = 25;
            auto time_elapsed = std::chrono::duration_cast<std::chrono::minutes>(now - last_updated);
            double new_points = std::min(max_points, (time_elapsed.count() / 15.0) * max_points);
            new_points = std::round(new_points * 100.0) / 100.0;

            std::random_device rd;
            std::mt19937 gen(rd());
            std::uniform_real_distribution<> dis(0.0, 1.0);
            if (dis(gen) < 0.1) {
                double bonus = dis(gen) * 2;
                new_points = std::min(max_points, new_points + bonus);
                new_points = std::round(new_points * 100.0) / 100.0;
            }

            potential_points = new_points;
        } else {
            countdown = "Calculating...";
            potential_points = 25.0;
        }
    } else {
        countdown = "Calculating...";
        potential_points = 0.0;
    }

    set_local_storage({{"potentialPoints", potential_points}, {"countdown", countdown}});
}

void start_countdown_and_points() {
    countdown_thread = std::thread([&]() {
        while (true) {
            update_countdown_and_points();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }
    });
}

std::string get_user_id() {
    std::string login_url = "https://ikknngrgxuxgjhplbpey.supabase.co/auth/v1/token?grant_type=password";
    std::string authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";
    std::string apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";

    std::string email, password;
    std::cout << "Email: ";
    std::cin >> email;
    std::cout << "Password: ";
    std::cin >> password;

    CURL *curl = curl_easy_init();
    if (curl) {
        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, ("Authorization: " + authorization).c_str());
        headers = curl_slist_append(headers, ("apikey: " + apikey).c_str());
        headers = curl_slist_append(headers, "Content-Type: application/json");

        json post_data = {
            {"email", email},
            {"password", password}
        };
        std::string post_fields = post_data.dump();

        curl_easy_setopt(curl, CURLOPT_URL, login_url.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, post_fields.c_str());

        std::string response;
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, [](void *contents, size_t size, size_t nmemb, std::string *s) -> size_t {
            size_t newLength = size * nmemb;
            s->append((char*)contents, newLength);
            return newLength;
        });
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            std::cerr << "curl_easy_perform() failed: " << curl_easy_strerror(res) << std::endl;
        } else {
            json response_json = json::parse(response);
            std::string user_id = response_json["user"]["id"];
            std::cout << "User ID: " << user_id << std::endl;

            std::string profile_url = "https://ikknngrgxuxgjhplbpey.supabase.co/rest/v1/profiles?select=personal_code&id=eq." + user_id;
            curl_easy_setopt(curl, CURLOPT_URL, profile_url.c_str());
            curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);

            response.clear();
            res = curl_easy_perform(curl);
            if (res != CURLE_OK) {
                std::cerr << "curl_easy_perform() failed: " << curl_easy_strerror(res) << std::endl;
            } else {
                json profile_data = json::parse(response);
                std::cout << "Profile Data: " << profile_data.dump(4) << std::endl;
            }

            set_local_storage({{"userId", user_id}});
            return user_id;
        }

        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
    }

    return "";
}

int main() {
    json local_storage_data = get_local_storage();
    std::string user_id = local_storage_data.value("userId", "");

    if (user_id.empty()) {
        std::cout << "Please enter your user ID: ";
        std::cin >> user_id;
        set_local_storage({{"userId", user_id}});
    }

    std::cout << "Menu:\n1. Logout\n2. Start Running Node\n3. Get User ID\nChoose an option: ";
    int option;
    std::cin >> option;

    switch (option) {
        case 1:
            set_local_storage(json::object());
            std::cout << "Logged out successfully." << std::endl;
            break;
        case 2:
            start_countdown_and_points();
            connect_websocket(user_id);
            break;
        case 3:
            user_id = get_user_id();
            break;
        default:
            std::cout << "Invalid option. Exiting..." << std::endl;
            return 0;
    }

    if (countdown_thread.joinable()) {
        countdown_thread.join();
    }

    return 0;
}

